hermes-client/src/main/java/pl/allegro/tech/hermes/client/ReactiveHermesClient.java

```diff
@@ -0,0 +1,324 @@
+        currentlySending.incrementAndGet();
+        Mono<Result> sendOnceResult = sendOnce(message)
+                .flatMap(this::testRetryCondition)
+                .onErrorResume(this::mapExceptionToFailedAttempt)
+                .subscribeOn(scheduler);
+
+        return retry(message, sendOnceResult)
+                .doOnSuccess(hr -> {
+                    if (hr.response.isSuccess()) {
+                        handleSuccessfulRetry(message, hr.getAttempt());
+                    } else {
+                        handleFailure(message, hr.getAttempt());
+                    }
+                })
+                .map(result -> result.response)
+                .doFinally(s -> currentlySending.decrementAndGet());
+    }
+
+    private Mono<Attempt> retry(HermesMessage message, Mono<Result> sendOnceResult) {
+        Retry retrySpec = prepareRetrySpec(message);
+        return sendOnceResult
+                .flatMap(this::unwrapFailedAttemptAsException)
+                .retryWhen(retrySpec)
+                .subscriberContext(ctx -> ctx.put(RETRY_CONTEXT_KEY, HermesRetryContext.emptyRetryContext()))
+                .onErrorResume(Exception.class, exception -> {
+                    if (isRetryExhausted(exception)) {
+                        RetryFailedException rfe = (RetryFailedException) (exception.getCause());
+                        handleMaxRetriesExceeded(message, rfe.attempt);
+                        Throwable cause = rfe.getCause();
+                        if (cause instanceof ShouldRetryException) {
+                            ShouldRetryException sre = (ShouldRetryException) cause;
+                            return Mono.just((Attempt) Result.attempt(sre.hermesResponse, rfe.attempt));
+                        }
+                        handleFailure(message, rfe.getAttempt());
+                    }
+                    return Mono.error(exception);
+                });
+    }
+
+    private Mono<Attempt> unwrapFailedAttemptAsException(Result result) {
+        if (result instanceof Failed) {
+            Failed failed = (Failed) result;
+            return Mono.error(new RetryFailedException(failed.attempt, failed.cause));
+        } else {
+            return Mono.just((Attempt) result);
+        }
+    }
+
+    private Mono<Result> mapExceptionToFailedAttempt(Throwable throwable) {
+        return getNextAttempt().map(attempt -> Result.failure(throwable, attempt));
+    }
+
+    private Mono<Result> testRetryCondition(HermesResponse response) {
+        return getNextAttempt()
+                .map(attempt -> {
+                    if (retryCondition.test(response)) {
+                        return Result.failure(new ShouldRetryException(response), attempt);
+                    } else {
+                        return Result.attempt(response, attempt);
+                    }
+                });
+    }
+
+    private Retry prepareRetrySpec(HermesMessage message) {
+        if (!retrySleep.isZero()) {
```

Comment:
Just a detail, but this would be more readable if condition was reversed

Reply:
I believe it's a bug and no negation is necessary, in the first place. Please take a look at the code in the else branch:
```java
return Retry.backoff(maxRetries, retrySleep);
```
Since it's the else branch, then we already know that retrySleep is zero, and the code could be replaced with:
```java
return Retry.backoff(maxRetries, Duration.ZERO);
```
Moreover, I switched the condition and no test failed so I'm going to add some tests for this scenario.

Reply:
Great, nice catch :)